IGNITE-28587 Expand and sweep RELEASE_NOTES.txt (#13060)
apache/ignite: RELEASE_NOTES.txt

* Performance statistics.
* System views.
* Tracing instrumentation.
* .Net platform cache.
* СDC.
* Service Middleware.
* JDBC driver partition awareness.
* IndexQuery API.
* DumpReader consumer and DumpReaderConfiguration.
* Metric registry filter based on regular expression.
* System properties:
    IGNITE_WAIT_FOR_BACKUPS_ON_SHUTDOWN
    IGNITE_ENABLE_SUBQUERY_REWRITE_OPTIMIZATION
    IGNITE_SNAPSHOT_SEQUENTIAL_WRITE
    IGNITE_CONFIGURATION_VIEW_PACKAGES
* Control utility commands:
    Print metadata
    Print list of binary metadata types
    Print detailed info about specified binary type
    Removed the metadata of the specified type
    Update cluster metadata from specified file
    Print tracing configuration
    Reset all specific tracing configuration
    Set new tracing configuration

Ignite:
* Added SessionContext API.
* Added store crash recovery data to checkpoint recovery files.
* Added support for multiple storage paths per cache.
* Added TCP connection pool metrics.
* Added support for partition loss detection after cluster inactivity.
* Added system views recording to perfStat report.
* Added cache storage support for --persistence commands.

> * Added WAL state/enable/disable commands to the control utility.

Reviewer: Move to `control utility` section

> * Added checkpoint command to control utility.

Reviewer: Move to `control utility` section

* Added cache names filter to dump reader.
* Added metrics to improve node network unavailability detection.
* Added authorization for Compute Job cancellation.

> * Added offline command support to the control utility.

Reviewer: Move to `control utility` section

* Fixed hangs in put operations with RANDOM_LRU and RANDOM_2_LRU eviction policies.
* Fixed read-through skipping in DML operations.
* Fixed an issue where IndexQuery couldn't use an index until data population completed.
* Fixed a memory leak in HeavyQueriesTracker.
* Fixed freelist corruption in JDK 17 with Shenandoah GC and assertions turned off.
* Fixed query entity configuration merge on node join.
* Fixed NullPointerException when getting the cluster name on an inactive cluster.
* Fixed CdcCommandLineStartup interruption.
* Fixed snapshot restore failures due to non-serializable cache group names collections.
* Fixed a binary compatibility issue with the dcId property.
* Fixed transaction closing on explicit rollback or commit.
* Fixed snapshot restoration when indexPath was equal to storagePath.
* Fixed reading of empty unsupported system views in perfStat.
* Fixed inconsistent behavior on concurrent commit and timeout with 1PC.
* Fixed suboptimal expire shift calculation.
* Fixed the corrupted tree exception on expiration.
* Fixed primitive column descriptor problem with alias on key or val field.
* Fixed H2QueryInfo#planWithoutScanCount for UNION queries.

> * Fixed primitive column descriptor problem with alias on key or val field.

Reviewer: Duplicate?

* Optimized discovery for multi-DC scenarios.
* Optimized rebalancing in multi-DC mode.
* Optimized the Cache API for multi-DC.
* Optimized SQL Calcite-related operations for multi-DC.
* Optimized the Java thin client for multi-DC scenarios.

> * Speed up warmup running in several threads.

Reviewer: Optimized cache warmup using multithreading?

> * Make "metastorage" and "distributedMetastorage" system views filtrable.

Reviewer: Added filtering capabilities to ...?

> * CdcMain clears corrupted cdc state files.

Reviewer: Fixed CdcMain to clear ...?

* Updated jetty-server dependency version to 11.0.25.
* Updated commons-lang3 dependency version to 3.18.0.

SQL:

> * Added a system view and fixed a performance issue in SQL query plan history. The default H2 SQL plan history was set to zero.

Reviewer: `Added a system view for SQL query plan history.` is enough

(!) WARNINGS:
* Removed deprecated GridClient.
* Removed deprecated SpiQuery.

* SqlFieldsQuery#setLazy/isLazy marked as deprecated.
* DeploymentSPI marked as deprecated.

No longer experimental parts and components:
* Calcite-based query engine.
* Performance statistics.
* System views.
* Tracing instrumentation.
* .Net platform cache.
* СDC.
* Service Middleware.
* JDBC driver partition awareness.
* IndexQuery API.
* DumpReader consumer and DumpReaderConfiguration.
* Metric registry filter based on regular expression.
* System properties:
    IGNITE_WAIT_FOR_BACKUPS_ON_SHUTDOWN
    IGNITE_ENABLE_SUBQUERY_REWRITE_OPTIMIZATION
    IGNITE_SNAPSHOT_SEQUENTIAL_WRITE
    IGNITE_CONFIGURATION_VIEW_PACKAGES
* Control utility commands:
    Print metadata
    Print list of binary metadata types
    Print detailed info about specified binary type
    Removed the metadata of the specified type
    Update cluster metadata from specified file
    Print tracing configuration
    Reset all specific tracing configuration
    Set new tracing configuration

Ignite:
* Added SessionContext API.
* Added store crash recovery data to checkpoint recovery files.

> * Added JMH benchmark and helper for TPC-H queries.

Reviewer: This is also internal part, not even exposed to release package

> * Added multiple storage paths per cache.

Reviewer: configuration for ...

* Added metrics of TCP communication connection pool.
* Added support of partition loss detection after cluster inactivity.
* Added recording of systemViews to perfStat report.

> * Added per cache custom storagePath configuration.

Reviewer: The same as `multiple storage paths per cache`

> * Added max value aggregated interval metric.

Reviewer: It's an internal part, not visible to users. Flag "Release notes required" is cleared for this ticket. Why it's in the release notes list? All tickets without this flag should be excluded.

* Added support of cache storage for --persistence commands.

> * Added DataStorageConfiguration#extraSnapshotPaths.

Reviewer: This all should be added under one epic `multiple storage paths`. Why do we add each minor ticket?

* Fixed in-memory cluster hanging when error is thrown on activation.

> * Optimized discovery in multi-DC scenarios.

Reviewer: I think all multi-DC tickets should be grouped under one epic, or at least located together

* Fixed hang in put with RANDOM_LRU and RANDOM_2_LRU eviction policy.

> * Added WAL state/enable/disable commands.

Reviewer: To control utility?

> * Skip read-through in DML operations.

Reviewer: Fixed skipping read-through in DML operations

> * Added an affinity backup filter assigning partition copies evenly between data centers.

Reviewer: MultiDC

> * Added checkpoint command.

Reviewer: Control utility?

* Fixed IndexQuery can't use an index until data population completes.

> * Removed setSharedMemoryPort.

Reviewer: Move to warnings?

* Added cache names filter to dump reader.
* Fixed memory leak in HeavyQueriesTracker.
* Added metrics to improve node network unavailability detection.
* Added authorization of Compute Job cancellation.

> * Added offline commands support.

Reviewer: Control utility?

* Added metrics for inserted/removed bytes to/from cache group.

> * Optimized rebalancing in multi-DC mode.

Reviewer: MultiDC epic

* Fixed freelist corruption in JDK17 with ShenandoahGC and assertions turned off.
* Added capturing of cluster node certificates during join process.
* Fixed query entity configuration merge on node join.
* Fixed NPE when getting cluster name on inactive cluster.
* Fixed CdcCommandLineStartup interruption.

> * Use MessageSerializer for SchemaOperationStatusMessage.

Reviewer: Internal

* CdcMain clears corrupted cdc state files.
* Fixed transaction closing on explicit rollback or commit.
* Added tx deadlocks count metric.

> * Removed redundant IgniteExperimental annotations.

Reviewer: It's already mentioned

* Fixed snapshot restoration when indexPath equal to storagePath.

> * Cache API optimizations for MultiDC.

Reviewer: MultiDC epic

* Added ability to create tables for already existing cache with defined schema.
* Fixed reading empty unsupported system views in perfStat.
* Fixed inconsistent behaviour on concurrent commit and timeout with 1PC.
* Fixed suboptimal expire shift calculation.
* Fixed corrupted tree exception on expiration.
* Speed up warmup running in several threads.
* Added performance statistics for putAllConflict, removeAllConflict operations.

> * Updated jetty-server dependency version.

Reviewer: to version ...

* Fixed primitive column descriptor problem with alias on key or val field.
* Make "metastorage" and "distributedMetastorage" system views filtrable.
* Added warn if index inline size is too big.
* Fixed H2QueryInfo#planWithoutScanCount for UNION queries.
* Documentation of distributed properties for blocking new connections.
* Added LastArchivedSegment metric.
* Fixed snapshot restore failure due to non-serializable cache group names collection.
* Fixed binary compatibility issue with dcId property.
* Added CODEOWNERS for platforms.

> * Added data center ID to the ClusterNodeView.

Reviewer: MultiDC epic

> * Updated dependency version commons-lang3.

Reviewer: to version

> * Added cache storage path to ClientCacheConfiguration.

Reviewer: multi storage epic

Reviewer: I think this item can be reomved, it's a part of`Added support for multiple storage paths per cache.`

* Added client router node metric.

SQL:
* Added system view for SQL query plans history.

> * Added docs for PostgreSQL cdc-extension.

Reviewer: Not sure docs should be mentioned in release notes

> * Fixed performance drop introduced by SQL plan history system view.

Reviewer: This was caused by `Added system view for SQL query plans history`. It shouldn't be in one release notes `added feature` and `fixed bug in this feature`.

> * Added thin client API to create SQL caches.

Reviewer: Related to thin clients

* Fixed node crash after SQL query with WITH clause.
* Added initiatorId to the SQL_QUERY_HISTORY system view.
* Fixed indexed field validation.
* Added information about query initiator id to log.

> * Set H2 default SQL plan history to zero.

Reviewer: Related to `Added system view for SQL query plans history`

Reviewer: Not removed

SQL Calcite:
* Updated Calcite version to 1.40.
* Added UNNEST operator support.
* Added HASH join type.
* Added query blocking tasks executor (allows to execute SQL inside UDF).
* Added non-equi condition support for LEFT/RIGHT/FULL/ANTI HASH JOIN.
* Added heuristics to optimize join order.
* Fixed MergeJoin/NestedLoopJoin buffers overflow.
* Added support of user-defined table functions.

> * Optimize scans with filter.

Reviewer: Optimized?

* Added dynamic configuration to disable a rule globally.
* Fixed merge filter to scan after trimmed fields.
* Added thread pool starvation warning, add UDF query warning.

> * Optimize memory consumption for hash-table based operators.

Reviewer: Optimized

> * Optimize IN statement with literals.

Reviewer: Optimized

> * Optimization for MultiDC.

Reviewer: MultiDC epic

* Added number of passed query parameters validation.
* Fixed incorrect processing of IS NOT DISTINCT condition in MergeJoin.

> * Limit joins number to commute in planning.

Reviewer: Limited. Related to `Added heuristics to optimize join order`

> * Fixed usage of checkState() in the join nodes.

Reviewer: Internal

* Fixed cast of dynamic parameters.
* Fixed wider type for fixed chars coercion.
* Fixed NVL/COALESCE failure with correlated subquery as parameter.
* Fixed excessive log output on planning error.

> * Reduced contention by QueryBlockingTaskExecutor.

Reviewer: Related to `Added query blocking tasks executor`, no additional mention required.

* Fixed assertion on planning for filter with double correlate.

> * Reuse Row for non-storing accumulators.

Reviewer: Something like 'Optimized row reusing`?

* Fixed function with schema serialization.
* Fixed "field names not distinct" assertion error.

> * Fixed flaky CancelTest.testNotOriginatorNodeStop.

Reviewer: Internal

Control utility:

> * Added support for interactive input of sensitive command arguments in control utility.

Reviewer: We are in `control utility` section, no need to mention `control utility` and `control.sh` in each item.

* Added rolling upgrade commands for control.sh.

> * Implement control.sh status command for rolling upgrade.

Reviewer: Added. Also looks like status is a part of `Added rolling upgrade commands` mentioned earlier

* Fixed control.sh --kill compute command authorization and authorization error propagation.
* Removed redundant permission check to cancel services from control utility.

Java thin client:

> * SqlFieldsQuery initiator ID support.

Reviewer: Added

> * Fixed NPE on SQL fields query.

Reviewer: Related to `SqlFieldsQuery initiator ID support` (once again, ticket without `Release notes required`)

* Fixed multi-key operations unordered map/set warnings in server logs.

> * Optimizations for MultiDC.

Reviewer: MultiDC epic

* Fixed thin client duplicates channels.
* Added handshake timeout configuration for java thin client.
* Fixed management client connection if the 'newThinConnectionsEnabled' property is set.
* Fixed binary meta sending on reconnect.

> * Fixed cache list command if mapped to a client node.

Reviewer: Looks like this is related to control utility

-----------------------------------------------------------
(!) WARNINGS:
* Removed deprecated GridClient.
* Removed deprecated SpiQuery.

* SqlFieldsQuery#setLazy/isLazy marked as deprecated.
* DeploymentSPI marked as deprecated.

No longer experimental parts and components:
* Calcite-based query engine.
* Performance statistics.
* System views.
* Tracing instrumentation.
* .Net platform cache.
* СDC.
* Service Middleware.
* JDBC driver partition awareness.
* IndexQuery API.
* DumpReader consumer and DumpReaderConfiguration.
* Metric registry filter based on regular expression.
* System properties:
    IGNITE_WAIT_FOR_BACKUPS_ON_SHUTDOWN
    IGNITE_ENABLE_SUBQUERY_REWRITE_OPTIMIZATION
    IGNITE_SNAPSHOT_SEQUENTIAL_WRITE
    IGNITE_CONFIGURATION_VIEW_PACKAGES
* Control utility commands:
    Print metadata
    Print list of binary metadata types
    Print detailed info about specified binary type
    Removed the metadata of the specified type
    Update cluster metadata from specified file
    Print tracing configuration
    Reset all specific tracing configuration
    Set new tracing configuration

Ignite:
* Added SessionContext API.

> * Store crash recovery data to checkpoint recovery files.

Reviewer 2: ```suggestion
* Added store crash recovery data to checkpoint recovery files.
```

Author: done

* Added JMH benchmark and helper for TPC-H queries.

> * Several storage paths per cache.

Reviewer 2: ```suggestion
* Added multiple storage paths per cache.
```

Author: done

* Added metrics of TCP communication connection pool.
* Added support of partition loss detection after cluster inactivity.

> * Recording systemViews to perfStart report.

Reviewer 2: ```suggestion
* Added recording of systemViews to perfStat report.
```

Author: done

> * Per cache custom storagePath configuration.

Reviewer 2: ```suggestion
* Added per cache custom storagePath configuration.
```

* Added max value aggregated interval metric.

> * Support of cache storage for --persitence commands.

Reviewer 2: ```suggestion
* Added support of cache storage for --persistence commands.
```

> * DataStorageConfiguration#extraSnapshotPaths added.

Reviewer 2: ```suggestion
* Added DataStorageConfiguration#extraSnapshotPaths.
```

> * Ignite accepts joining node with rolling upgrade version.

Reviewer 2: Is rolling upgrade fully supported? This looks like an internal detail of a bigger feature.

* Fixed in-memory cluster hanging when error is thrown on activation.

> * Discovery optimizations for MultiDC.

Reviewer 2: ```suggestion
* Optimized discovery in multi-DC scenarios.
```

* Fixed hang in put with RANDOM_LRU and RANDOM_2_LRU eviction policy.

> * WAL state/enable/disable commands.

Reviewer 2: ```suggestion
* Added WAL state/enable/disable commands.
```

* Skip read-through in DML operations.

> * Implement an affinity backup filter assigning partition copies evenly between data centers.

Reviewer 2: ```suggestion
* Added an affinity backup filter assigning partition copies evenly between data centers.
```

> * Сheckpoint command added.

Reviewer 2: ```suggestion
* Added checkpoint command.
```

Added where?

> * IndexQuery can't use an index until data population completes.

Reviewer 2: Sounds like a bug description, was that fixed or?

* Removed setSharedMemoryPort.

> * Introduce CacheIdleVerifyCancelCommand.

Reviewer 2: Is this public API or something internal?

> * Compatibility test for Ignite snapshot.

Reviewer 2: ```suggestion
```
I don't think we need to mention tests in RN

> * Dump reader - cache names filter added.

Reviewer 2: ```suggestion
* Added cache names filter to dump reader.
```

* Fixed memory leak in HeavyQueriesTracker.
* Added metrics to improve node network unavailability detection.
* Added authorization of Compute Job cancellation.
* Added offline commands support.
* Added metrics for inserted/removed bytes to/from cache group.

> * Rebalancing optimization for MultiDC.

Reviewer 2: ```suggestion
* Optimized rebalancing in multi-DC mode.
```

* Fixed freelist corruption in JDK17 with ShenandoahGC and assertions turned off.
* Added capturing of cluster node certificates during join process.
* Fixed query entity configuration merge on node join.
* Fixed NPE when getting cluster name on inactive cluster.
* Fixed CdcCommandLineStartup interruption.
* Use MessageSerializer for SchemaOperationStatusMessage.
* CdcMain clears corrupted cdc state files.
* Fixed transaction closing on explicit rollback or commit.
* Added tx deadlocks count metric.
* Removed redundant IgniteExperimental annotations.
* Fixed snapshot restoration when indexPath equal to storagePath.
* Cache API optimizations for MultiDC.

> * Allow to create table for already existing cache with defined schema.

Reviewer 2: ```suggestion
* Added ability to create tables for already existing cache with defined schema.
```

> * PerfStat: fix reading empty unsupported system views.

Reviewer 2: ```suggestion
* Fixed reading empty unsupported system views in perfStat.
```

* Fixed inconsistent behaviour on concurrent commit and timeout with 1PC.

> * Disable distributed processes during rolling upgrade.

Reviewer 2: What does it mean for the user?

* Fixed suboptimal expire shift calculation.
* Fixed corrupted tree exception on expiration.
* Speed up warmup running in several threads.
* Added performance statistics for putAllConflict, removeAllConflict operations.
* Updated jetty-server dependency version.

> * Reduce memory requirements of largeUserAttributes* tests.

Reviewer 2: ```suggestion
```
Tests

* Fixed primitive column descriptor problem with alias on key or val field.
* Make "metastorage" and "distributedMetastorage" system views filtrable.
* Added warn if index inline size is too big.
* Fixed H2QueryInfo#planWithoutScanCount for UNION queries.
* Documentation of distributed properties for blocking new connections.
* Added LastArchivedSegment metric.
* Fixed snapshot restore failure due to non-serializable cache group names collection.
* Fixed binary compatibility issue with dcId property.
* Added CODEOWNERS for platforms.
* Added data center ID to the ClusterNodeView.
* Updated dependency version commons-lang3.

> * Thick client optimizations for MultiDC.

Reviewer 2: ```suggestion
```

MultiDC optimizations already mentioned above

> * Cache storage path added to ClientCacheConfiguration.

Reviewer 2: ```suggestion
* Added cache storage path to ClientCacheConfiguration.
```

* Added client router node metric.

SQL:
* Added system view for SQL query plans history.
* Added docs for PostgreSQL cdc-extension.
* Fixed performance drop introduced by SQL plan history system view.

> * Method to create SQL cache via thin client.

Reviewer 2: ```suggestion
* Added thin client API to create SQL caches.
```

* Fixed node crash after SQL query with WITH clause.
* Added initiatorId to the SQL_QUERY_HISTORY system view.
* Fixed indexed field validation.
* Added information about query initiator id to log.
* Set H2 default SQL plan history to zero.

> * Documentation: Add SQL_PLANS_HISTORY view performance considerations.

Reviewer 2: ```suggestion
```
We don't usually mention docs

SQL Calcite:

> * Update Calcite version to 1.40.

Reviewer 2: ```suggestion
* Updated Calcite version to 1.40.
```

> * Support UNNEST operator.

Reviewer 2: ```suggestion
* Added UNNEST operator support.
```

* Added HASH join type.
* Added query blocking tasks executor (allows to execute SQL inside UDF).
* Added non-equi condition support for LEFT/RIGHT/FULL/ANTI HASH JOIN.
* Added heuristics to optimize join order.
* Fixed MergeJoin/NestedLoopJoin buffers overflow.
* Added support of user-defined table functions.
* Optimize scans with filter.
* Added dynamic configuration to disable a rule globally.
* Fixed merge filter to scan after trimmed fields.
* Added thread pool starvation warning, add UDF query warning.
* Optimize memory consumption for hash-table based operators.
* Optimize IN statement with literals.
* Optimization for MultiDC.
* Added number of passed query parameters validation.
* Fixed incorrect processing of IS NOT DISTINCT condition in MergeJoin.
* Limit joins number to commute in planning.
* Fixed usage of checkState() in the join nodes.
* Fixed cast of dynamic parameters.
* Fixed wider type for fixed chars coercion.
* Fixed NVL/COALESCE failure with correlated subquery as parameter.
* Fixed excessive log output on planning error.

> * Reduce contention by QueryBlockingTaskExecutor.

Reviewer 2: ```suggestion
* Reduced contention by QueryBlockingTaskExecutor.
```
